### PR TITLE
[SPARK-36663][SQL] Support number-only column names in ORC data sources

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR aims to support number-only column names in ORC data sources.
In the current master, with ORC datasource, we can write a DataFrame which contains such columns into ORC files.
```
spark.sql("SELECT 'a' as `1`").write.orc("/path/to/dest")
```

But reading the ORC files will fail.
```
spark.read.orc("/path/to/dest")
...
== SQL ==
struct<1:string>
-------^^^

  at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:265)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:126)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parseDataType(ParseDriver.scala:40)
  at org.apache.spark.sql.execution.datasources.orc.OrcUtils$.org$apache$spark$sql$execution$datasources$orc$OrcUtils$$toCatalystSchema(OrcUtils.scala:91)
```
The cause of this is `OrcUtils.toCatalystSchema` fails to handle such column names.
In `OrcUtils.toCatalystSchema`, `CatalystSqlParser` is used to create a instance of `StructType` which represents a schema but `CatalystSqlParser.parseDataType` fails to parse if a column name (and nested field) consists of only numbers.

### Why are the changes needed?

For better usability.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New tests.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @sarutak . AFAIK, this can be a feature parity issue between Parquet and ORC, but this isn't a regression in ORC support in Apache Spark. IIRC, we never support this before in the official releases including 2.4.8/3.0.2/3.1.2/3.2.0(RC2).

I revised the PR title and JIRA issue type as `Improvement` for Apache Spark 3.3.0. This is not a blocker for Apache Spark 3.2.0 RC or a bug in previous released versions.

**Comment by sarutak** [comment]
@dongjoon-hyun Thank you for taking a look at this. I agree that this is an improvement rather than a bug fix.

**Comment by HyukjinKwon** [comment]
I think it looks fine ... but would better to have Dongjoon's sign-off.

**Comment by sarutak** [comment]
Yeah, I'll wait for his review.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:95]: I see, now we only parse the leaf type name, and don't need to parse the struct field name. +1

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:102]: This is not efficient as we create many StructType instances. Can we just create StructFields and then build `StructType` only once?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:105]: does orc struct have nullability and metadata?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:110]:
```suggestion
      val elementType = orcType.getChildren(0)
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (sarutak) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:105]: All the types in ORC seems to have nullability.
https://orc.apache.org/docs/types.html

Schema mapping from Catalyst to ORC seems to be done using `OrcFileFormat.getQuotedSchemaString`, and it doesn't consider only `name` and `dataType` so I think we don't need to care about metadata here.

https://github.com/apache/spark/blob/ff7705ad2ad5f4b9dfbeda83e93a0db676e1ffd9/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFileFormat.scala#L47-L59

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:110]: then `.get(0)`?

Comment (sarutak) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala:110]: Exactly. I have no idea why I didn't notice that...

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**
